Reorder items in View menu

gerhardol opened the pull request:
## Proposed changes

Reorder the menu items to reflect how the items are used
This menu has grown without much consideration over the years. The sorting options were recently changed in #10172 to be decoupled.
Changes the muscle memory, but even I have a problem understanding the type of setting today...

* Branch filter mode (reverse order to how they are prioritized) 
Reflog <- current branch <- filtered branches <- All
(Except for reflog visible in Advanced filter)
Note: Any branch filter could be ticked at the same time as Reflog, previously
Reflog added to the branch filter dropdown.


* Further revision filters, mostly independent to filter modes 
Show merges, first parents
(Other revision filters in Advanced filter)

* History simplification, how to handle history with filters
Full history, simplify merges.
(Simplify by decoration in Advanced filter handled as a revision filter despite Git description https://git-scm.com/docs/git-log#Documentation/git-log.txt---simplify-by-decoration)

* Highlighting etc

* Added commits in grid
Artificial commits, stash, Git notes

* Labels in Grid 
Remote branches, tags, superproject info

* Specific details in the grid

* Columns to show

* Sorting

* Advanced filter form

~Submenus could be added to explain the groups, should be a separate PR~
Added group headers to explain usage

This will require an update in the doc too (a few other changes needed as well)
https://git-extensions-documentation.readthedocs.io/en/release-4.0/browse_repository.html

There is some overlap between the Advanced filter and View menu.
All revision filters and simplify history should be moved there as well as the reflog *branch revision filter option* be added.
That should be a separate PR

## Screenshots <!-- Remove this section if PR does not change UI -->

| Before | After |
|---|---|
| ![image](https://user-images.githubusercontent.com/6248932/189451145-08deaa7e-506a-4b69-b70f-766c2f32d28f.png) | ![image](https://user-images.githubusercontent.com/6248932/190252116-e679780f-396c-4750-8a5c-a76ad9e6244a.png) |
| ![image](https://user-images.githubusercontent.com/6248932/189500626-37a91740-4e3d-44e1-91db-7770856023d4.png) | ![image](https://user-images.githubusercontent.com/6248932/189551086-7b3a74fd-3cf5-4708-a1b6-10a91269f80d.png)  |

(Before displays _Show stashes_ instead of _Show latest stashes_ as it is from a branch, no other changes.)

## Test methodology <!-- How did you ensure quality? -->

manual

## Merge strategy

I agree that the maintainer squash merge this PR (if the commit message is clear).

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

gerhardol commented:
Cleaning up the filterhandling slightly in addition to the menu (a few clarifications too)
Updated description and screenshots and listed some follow ups

mstv commented:
👍

Perhaps we should move down options which are configured once or very rarely. I.e. move up the others - which might be changed forth and back in different usecases - resulting in e.g.:
- Branch filter mode
- Advanced filter form
- Sorting
- Highlighting etc
  - Highlight selected branch
  - Draw non-relatives gray
- Show ...

Although the git argument is `--first-parents`, we should be more explicative: `Show first parents only`.

> Reflog should probably be added to the branch filter dropdown.

I agree.

mstv commented:
> Submenus could be added to explain the groups, should be a separate PR.

Only if it does not fit on the screen, which is not the case. It is hard to find items in submenus.

> There is some overlap between the Advanced filter and View menu.
All revision filters and simplify history should be moved there as well as the reflog branch revision filter option be added.
That should be a separate PR

(I think it is good to provide some of the options redundantly for quick access.)
Basically the Advanced Filter should become the superset of all filter options (in a separate PR).

![image](https://user-images.githubusercontent.com/36601201/189524077-9f5778c5-8254-447e-b513-a2d99e672c8b.png)

gerhardol commented:
My first priority with this PR is that it is included with next 4.0 public build.
@RussKie was to make a build next week, I want this included.
Other discussions in a follow up if they can wait.

> Although the git argument is --first-parents, we should be more explicative: Show first parents only.

OK, but next commit, since I want to remove that option from this menu and that will require the translation to be changed twice.

 >> Submenus could be added to explain the groups, should be a separate PR.

> Only if it does not fit on the screen, which is not the case. It is hard to find items in submenus.

It is a way to describe what the group is doing, so it could be adding value also if the items exist.

> (I think it is good to provide some of the options redundantly for quick access.)

I want to move the following to the Advanced filter form - they are not important and not really "View" either:
* *Show first parents* (button exists too)
* * Show merge commits* (Completely meaningless before this PR, quite meaningless after)
* Full history (not expected to change often)
* Simplify merges (dependent on the previous)

Branch revision filters are available in the toolbar as well as Advanced filter (except for reflog that is not in the form but in the toolbar even not in the dropdown).
They should probably be removed from View.

> Perhaps we should move down options which are configured once or very rarely. I.e. move up the others - which might be changed forth and back in different usecases - resulting in e.g.:

Suggested order with motivation. This is what I would like to settle the most right now.
Proposed submenu names in bold.

*  Branch filter revision
If they are in the menu (I suggest the first three proposed groups are removed in a followup), they should be first.

* Highlighting etc
        Highlight selected branch
        Draw non-relatives gray
OK to move this up, changing the PR (they will be on top)

* **Additional commits*
Artificial, stash, notes

* **Labels**
branches and tags

* **Columns**

* **Contents** (Better name needed...

*     Sorting
Often shown close to last in our menus

* Advanced filter form
A button on the toolbar and a shortcut exists

* **Toolbars** (as now)

mstv commented:
Regardless of what designers think, I _know_ that submenus give a bad UX to me. Aiming and hovering is OK for tooltips. Being forced to wait one second until info is faded in, is a real pain (e.g. Windows taskbar).
I understand your point regarding headings. Though they can be added directly:

![image](https://user-images.githubusercontent.com/36601201/189546646-9e30465f-f55a-4605-9dfc-f4b2263d2d80.png)

---

It stand by my argument regarding the logical linkage and usage frequency of the sections.

gerhardol commented:
> Regardless of what designers think, I _know_ that submenus give a bad UX to me. Aiming and hovering is OK for tooltips. Being forced to wait one second until info is faded in, is a real pain (e.g. Windows taskbar). I understand your point regarding headings. Though they can be added directly:

That is a hack, will consider to add to this PR still
 
> It stand by my argument regarding the logical linkage and usage frequency of the sections.

So do I...
Pushed moving Highlight further up (not updated the screenshot yet)

gerhardol commented:
Pushed changes after latest discussions with @mstv (not agreeing on the group order in View menu)
Follow up PR to move filter items to Advanced filter
(This was meant to be a quick fix...)

@RussKie @pmiossec 
I would like to include this in next 4.0 build

RussKie commented:
I'm currently away, but I'll try to review this in the next few days.

On Mon, 12 Sept 2022, 8:18 am Gerhard Olsson, ***@***.***>
wrote:

> Pushed changes after latest discussions with @mstv
> <https://github.com/mstv> (not agreeing on the group order in View menu)
> Follow up PR to move filter items to Advanced filter
> (This was meant to be a quick fix...)
>
> @RussKie <https://github.com/RussKie> @pmiossec
> <https://github.com/pmiossec>
> I would like to include this in next 4.0 build
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/10189#issuecomment-1243054234>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABBTEXVIBYTP4J7C4LLGFNDV5ZLDFANCNFSM6AAAAAAQJBF3HY>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

RussKie reviewed (commented):
I like what I see on the screenshots. Assuming these are current - 👍

pmiossec commented:
I'm OK to have it included in v4.0.
I'm testing it and it's good from my first uses.
I don't have a strong opinion on the sorting, I nearly never go to this menu, except very rarely for one of the first entries...

The disabled titles are a good idea but I wonder if we should not add a colon at the end to make it even clearer that it's not a menu entry that you could enable it one way or the other.

Another remark on the View menu but I think it will surely be for later (because maybe it won't be possible for all the options, for example "Commit message body"): I like how the menuitems in the "Toolbars" submenu are handled where it doesn't close the View menu and you could see it applied in "live". I would love to have the same behavior for most of the "Grid Info" and "Columns" sections menuitems. It will make the use of these options much easier especially for beginners when you don't understand the effects and have to try and revert...

A consequence to get consistency is that maybe theses sections should be put in submenus.

gerhardol commented:
> I like what I see on the screenshots. Assuming these are current

Yes, updated

> The disabled titles are a good idea but I wonder if we should not add a colon at the end to make it even clearer that it's not a menu entry that you could enable it one way or the other.

We do not have colons for other fake headings like in Copy, some in RevDiff
Not important for me, will add if someone agree

> Another remark on the View menu but I think it will surely be for later (because maybe it won't be possible for all the options, for example "Commit message body"): I like how the menuitems in the "Toolbars" submenu are handled where it doesn't close the View menu and you could see it applied in "live". I would love to have the same behavior for most of the "Grid Info" and "Columns" sections menuitems. It will make the use of these options much easier especially for beginners when you don't understand the effects and have to try and revert...

The problem here is that the grid need to be refreshed, the required data is not calculated if not needed. Maybe columns and a few more could be handled this way, most requires to read all data from scratch.

--

Will fix the failing test, as well allow toggling with the reflog button (now it resets the branch pattern to AllBranches).

RussKie commented:
> Another remark on the View menu but I think it will surely be for later (because maybe it won't be possible for all the options, for example "Commit message body"): I like how the menuitems in the "Toolbars" submenu are handled where it doesn't close the View menu and you could see it applied in "live". I would love to have the same behavior for most of the "Grid Info" and "Columns" sections menuitems. It will make the use of these options much easier especially for beginners when you don't understand the effects and have to try and revert...

I'd be advising against any UI/UX-related changes that deviate from the standard Windows UI/UX paradigms (e.g., this is not how Windows Explorer works). In cases where a live preview would be required, we should come up with different presentation mechanisms.

RussKie commented on GitUI/UserControls/RevisionGrid/RevisionGridControl.cs:
```diff
@@ -1724,6 +1730,20 @@ public void ShowFilteredBranches()
             PerformRefreshRevisions();
         }
 
+        public void ShowCurrentBranchOnly()
```
[nit] Can we please retain `ShowCurrentBranchOnly` where it was (i.e., lines:1690-...) and place `ShowReflogs` method here (i.e., after `ShowFilteredBranches`)?

gerhardol replied:
I reordered them to match how they are used, but can revert.

RussKie commented:
The menu is getting longer, and I think it's worth considering how we can separate the grid configuration related items (e.g., the column visibility) from the graph rendering. However, this is a good step. Thank you.

gerhardol commented:
> The menu is getting longer, and I think it's worth considering how we can separate the grid configuration related items (e.g., the column visibility) from the graph rendering. However, this is a good step. Thank you.

The first I would like to do is do move the four revision filter items to Advanced form, possibly the branch revision filter too. Then all revision filter handling can be moved to FilterInfo and some simplification and bug fixes (or "limitations") can be done. (Maybe even for 4.0.) https://github.com/gitextensions/gitextensions/pull/10189#issuecomment-1243027855
Then the size is manageable - for now.

But before merging I would like a mediation with @mstv comments in https://github.com/gitextensions/gitextensions/pull/10189#issuecomment-1242938595

gerhardol commented:
> Should the headings have an italic font style as in the other menus?

OK, implemented a further hack

> What do you think about "Show first parents **only**"?

Fine, but I want to move the option to Advanced form (keep the button).

>     * ordered by amount of displayed revisions
>     * ordered by ascending priority

I had changed to the first alternative, not updated the screenshot, done now

pmiossec commented:
> > ```
> > * ordered by amount of displayed revisions
> > * ordered by ascending priority
> > ```
> 
> 
> I had changed to the first alternative, not updated the screenshot, done now

That brings for me another question (sorry also): if there are in fact the same options, should we also add the same icons in the view menu?

gerhardol commented:
> > > ```
> > > * ordered by amount of displayed revisions
> > > * ordered by ascending priority
> > > ```
> > 
> > 
> > I had changed to the first alternative, not updated the screenshot, done now
> 
> That brings for me another question (sorry also): if there are in fact the same options, should we also add the same icons in the view menu?

The same options. Images could be the same, or the options removed
(better to ask before than after)

pmiossec commented:
> The same options. Images could be the same, or the options removed
> (better to ask before than after)

I prefer adding the icons. I imagine that some users are used to go in the view menus to toggle it (even if I find the toolbar more convenient). But let's see what the others think about it...

mstv reviewed (commented):
+1 after updating translation https://ci.appveyor.com/api/buildjobs/joyxw2k74hpbihte/artifacts/English.xlf

gerhardol commented:
Changed the "Group headers in italics" to use some more common code from an existing class
Added images to branch filters
Updated screenshot

Time to squash?

gerhardol commented:
Mockup of Advanced filter after options are moved there - future PR
I do not want to delay a release though...

![image](https://user-images.githubusercontent.com/6248932/190274872-597ceb94-2674-4ddf-997b-bb6728b0d319.png)

gerhardol commented on GitUI/UserControls/FilterToolBar.cs:
```diff
@@ -24,9 +24,11 @@ internal partial class FilterToolBar : ToolStripEx
         public FilterToolBar()
         {
             InitializeComponent();
+            tsmiShowReflogs.ToolTipText = TranslatedStrings.ShowReflog;
```
The Git command is "reflog" but that doc is stating *Reference logs, or "reflogs"*
Maybe add it instead?
Should be consistent
https://git-scm.com/docs/git-reflog

gerhardol replied:
Google for "git reflog" 2.6M and "git reflogs" 0.9M.
Changing to reflog (except for `fsck-objects --no-reflogs`)

Rebased on master to resolve conflicts

mstv commented:
> Time to squash?

Yes, please.

gerhardol commented:
> > Time to squash?
> 
> Yes, please.

Left a reference in tmp/reorder-view-menu

RussKie commented:
> Mockup of Advanced filter after options are moved there - future PR I do not want to delay a release though...
> 
> ![image](https://user-images.githubusercontent.com/6248932/190274872-597ceb94-2674-4ddf-997b-bb6728b0d319.png)

Any reason not to use PropertyGrid similar to how we configure Scripts?

gerhardol commented:
> Any reason not to use PropertyGrid similar to how we configure Scripts?

More work to restructure

gerhardol commented:
Merging this tomorrow

gerhardol commented:
My priority is not very high on such a change, there are more urgent changes. And I get very frustrated by UI changes, which makes my wife even less happy so I do not see I will change this in the near future.
